### PR TITLE
add build tests for the doc-build

### PR DESCRIPTION
**Pull request by syphar**
We don't have any tests for the build-process, here we can add some. 

I don't know the build-process in detail, so 
- What would be valid assertions here? 
- should we test `add-essential-files` separately? 
- what other tests should we have? 

My current (not yet validated) assumption is that the container and rustwide `.workspace` will only set up once per job, and not per test-case. 

Also, these tests would definitely fail on macs, and also on some linux-machines (cc @Nemo157 ). What would be the best rust-way of splitting them and only running them on CI? Not sure if `#[ignore]` and `cargo test -- --ignored` is too hidden

### CODE REVIEW

File: src/test/mod.rs [hunk 1]
```diff
@@ -137,8 +140,11 @@ impl TestEnvironment {
     fn base_config(&self) -> Config {
         let mut config = Config::from_env().expect("failed to get base config");
 
+        // create index directory
+        fs::create_dir_all(config.registry_index_path.clone()).unwrap();
+
         // Use less connections for each test compared to production.
-        config.max_pool_size = 2;
+        config.max_pool_size = 4;
```
Comment (jyn514) [inline src/test/mod.rs:147]: Why were these changes needed?

**Review by jyn514 — COMMENTED** [review]
> What would be valid assertions here?

See https://github.com/rust-lang/docs.rs/issues/822 for some suggestions.

> should we test add-essential-files separately?

I don't know how useful this is to test. It depends on the toolchain used, so rustdoc can break it at any time even if the tests pass in CI. I would rather have https://github.com/rust-lang/docs.rs/issues/1302.

> My current (not yet validated) assumption is that the container and rustwide .workspace will only set up once per job, and not per test-case.

Yes, it will take much too long otherwise. Even downloading `crates-build-env` takes a very long time, if there's a way to change that for tests it would be useful.

> What would be the best rust-way of splitting them and only running them on CI? Not sure if #[ignore] and cargo test -- --ignored is too hidden

This seems reasonable as a first step.

File: src/docbuilder/rustwide_builder.rs [hunk 2]
```diff
@@ -760,12 +760,75 @@ mod tests {
...
+            let base = format!("rustdoc/{}/{}", crate_, version);
+
+            // default target was build
+            assert!(storage.exists(&format!("{}/index.html", base))?);
+
+            // other targets too
+            for target in DEFAULT_TARGETS {
+                if target != &default_target {
+                    assert!(storage.exists(&format!("{}/{}/index.html", base, target))?);
+                }
```
Comment (jyn514) [inline src/docbuilder/rustwide_builder.rs:829]:
```suggestion
                let target_docs_present = storage.exists(&format!("{}/{}/index.html", base, target))?);
                if target != &default_target {
                    assert!(target_docs_present);
                } else {
                    assert!(!target_docs_present);
                }
```

File: src/docbuilder/rustwide_builder.rs [hunk 3]
```diff
@@ -760,12 +760,75 @@ mod tests {
...
+                    "x86_64-pc-windows-msvc",
+                ])
+            );
+
+            let storage = env.storage();
+            assert!(storage.exists("asdf")?);
+
+            let base = format!("rustdoc/{}/{}", crate_, version);
+
+            // default target was build
```
Comment (jyn514) [inline src/docbuilder/rustwide_builder.rs:822]:
```suggestion
            // default target was built
```

File: src/docbuilder/rustwide_builder.rs [hunk 4]
```diff
@@ -760,12 +760,75 @@ mod tests {
...
+            assert_eq!(row.get::<_, String>("default_target"), default_target);
+            assert_eq!(
+                row.get::<_, Value>("doc_targets"),
+                json!([
+                    "x86_64-unknown-linux-gnu", // default target is first
+                    "i686-pc-windows-msvc",     // rest is alphabetically
+                    "i686-unknown-linux-gnu",
+                    "x86_64-apple-darwin",
+                    "x86_64-pc-windows-msvc",
+                ])
```
Comment (jyn514) [inline src/docbuilder/rustwide_builder.rs:814]: Hmm, I don't know if we should guarantee the order. Maybe just sort the targets before adding the assert?

File: src/docbuilder/rustwide_builder.rs [hunk 5]
```diff
@@ -760,12 +760,75 @@ mod tests {
...
+            assert!(storage.exists(&format!("{}/index.html", base))?);
+
+            // other targets too
+            for target in DEFAULT_TARGETS {
+                if target != &default_target {
+                    assert!(storage.exists(&format!("{}/{}/index.html", base, target))?);
+                }
+            }
+
+            Ok(())
```
Comment (jyn514) [inline src/docbuilder/rustwide_builder.rs:832]: Can you add a test that hitting the website also works?

File: src/docbuilder/rustwide_builder.rs [hunk 6]
```diff
@@ -760,12 +760,75 @@ mod tests {
...
+            let default_target = "x86_64-unknown-linux-gnu";
+            // env.fake_release()
+            //     .name(crate_)
+            //     .version(version)
+            //     .default_target(default_target)
+            //     .add_target("i686-pc-windows-msvc")
+            //     .add_target("i686-unknown-linux-gnu")
+            //     .add_target("x86_64-apple-darwin")
+            //     .add_target("x86_64-pc-windows-msvc")
+            //     .create()?;
```
Comment (jyn514) [inline src/docbuilder/rustwide_builder.rs:778]: These aren't necessary because we're doing the real thing, not going through FakeRelease, right?

File: src/test/mod.rs (hunk 1 above)
Comment (syphar) [reply src/test/mod.rs:147]: the test was timing out to get a new connection

File: src/docbuilder/rustwide_builder.rs (hunk 6 above)
Comment (syphar) [reply src/docbuilder/rustwide_builder.rs:781]: these were remains of experiments :)

File: src/docbuilder/rustwide_builder.rs (hunk 4 above)
Comment (syphar) [reply src/docbuilder/rustwide_builder.rs:817]: The current impl seems to guarantee it, but I will change the test to sort first.

File: src/test/mod.rs (hunk 1 above)
Comment (jyn514) [reply src/test/mod.rs:147]: Ok, and creating the index directory?

Comment (syphar) [reply src/test/mod.rs:147]: also failed without, also this is needed for the local setup too ( as seen in the [readme](https://github.com/rust-lang/docs.rs#getting-started))

File: src/docbuilder/rustwide_builder.rs (hunk 4 above)
Comment (syphar) [reply src/docbuilder/rustwide_builder.rs]: actually, the tests failed on this :) So sorting it will be

**Comment by syphar** [comment]
@jyn514 this is ready for another round. 

Should we finish this up first? 

Or we can add more, coming from #551  I could see: 
- a build with a default target not being linux
- a build with more custom targets 
(for both it would help me to know crates that fix this description)

**Comment by jyn514** [comment]
> Should we finish this up first?

Landing an MVP sounds good to me. @pietroalbini I'd be interested to know what you think about this - it's closer to #524 than anything else, but it's better than not having tests at all. It does add an extra 10 minutes to CI unfortunately.

@syphar could you try switching the docker image to something smaller for CI? Maybe a base ubuntu:20.04 image? Right now I think `DOCS_RS_LOCAL_DOCKER_IMAGE` only supports local tags but it shouldn't be to hard to allow setting arbitrary remote images through an env variable.

**Comment by syphar** [comment]
> > Should we finish this up first?
> 
> Landing an MVP sounds good to me. @pietroalbini I'd be interested to know what you think about this - it's closer to #524 than anything else, but it's better than not having tests at all. It does add an extra 10 minutes to CI unfortunately.

I'm happy to add some more tests, since it works generally. 
Also, from my experience having a broken production system is worse than an extra 10 minutes in CI :) 


> @syphar could you try switching the docker image to something smaller for CI? Maybe a base ubuntu:20.04 image? Right now I think `DOCS_RS_LOCAL_DOCKER_IMAGE` only supports local tags but it shouldn't be to hard to allow setting arbitrary remote images through an env variable.

I'll give it a try (ignore the notifications until I ping you, I'll have to test with commits on github again)

**Comment by syphar** [comment]
@jyn514 short summary about the status quo: 
- using plain `ubuntu:20.04` didn't work because it doesn't have `build-essentials`. So I created a separate Dockerfile and used that to create the container to be used. 
- Also I change the test to use the dummy crate for testing. 
both these changes brought the time for the slow tests down from 9 minutes to around 4-5, while adding ~30 seconds for building the small image (caching the docker layers with a separate action makes it actually slower because of the other builds from `docker-compose`). 

Next improvements could be getting rid of the 2 min cargo-sweep install, but that I would look at separately.

**Comment by pietroalbini** [comment]
I don't have a chance to look into this more, but there are a couple of suggestions I have here:

* You should initialize Rustwide with [`fast_init(true)`](https://docs.rs/rustwide/0.12.0/rustwide/struct.WorkspaceBuilder.html#method.fast_init), which will speed up initialization.
* I'm pretty sure you can setup caching for the small Docker image using GitHub Actions.

**Review by jyn514 — APPROVED** [review]
I think we can address anything else in follow-ups :)

The time for this went down from 10 to 6 minutes after the smaller docker image and `fast_init(true)` - I'm not *super* happy it takes that long, but it's definitely better than not having tests at all.
